### PR TITLE
[pkg/pdatatest] replace compare of span event and link from reflect.DeepEqual to compare functions

### LINKED ISSUE #18049: [pkg/pdatatest] Expand comparison span links and events
Span events and links are currently compared with `reflect`. They need to be checked the same way as other data slices.

### PR DESCRIPTION
**Pull request by fatsheep9146**
Signed-off-by: Ziqi Zhao <zhaoziqi9146@gmail.com>

**Description:** 
fix #18049

### CODE REVIEW

File: .chloggen/span-links-events-compare.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
```
Comment (MovieStoreGuy) [inline .chloggen/span-links-events-compare.yaml:1]: I don't think a change log would needed for an implementation change.

**Review by MovieStoreGuy — APPROVED** [review]
I prefer this over using `reflect` since it is more descriptive on what isn't equal.

However, since there is no user impact or modification in behaviour then I believe it is fine to omit the changelog.

File: pkg/pdatatest/ptracetest/traces.go [hunk 2]
```diff
@@ -301,3 +297,84 @@ func CompareSpan(expected, actual ptrace.Span) error {
...
+	if expected.Len() != actual.Len() {
+		errs = multierr.Append(errs, fmt.Errorf("number of events doesn't match expected: %d, actual: %d",
+			expected.Len(), actual.Len()))
+		return errs
+	}
+
+	for i := 0; i < expected.Len(); i++ {
+		ee := expected.At(i)
+		ae := actual.At(i)
+		errs = multierr.Append(errs, internal.AddErrPrefix(fmt.Sprintf(`event "%s"`, ee.Name()), CompareEvent(ae, ee)))
```
Comment (dmitryax) [inline pkg/pdatatest/ptracetest/traces.go:313]: Can you please apply the same matching and out-of-order checks as done for other slices both for `SpanEventSlice` using `Name` as an identifier and `SpanLinkSlice` using `SpanID` as an identifier?

File: pkg/pdatatest/ptracetest/traces.go [hunk 3]
```diff
@@ -301,3 +297,84 @@ func CompareSpan(expected, actual ptrace.Span) error {
 
 	return errs
 }
+
+// CompareEvents compares each part of two given SpanEventSlice and returns
+// an error if they don't match. The error describes what didn't match.
+func CompareEvents(expected, actual ptrace.SpanEventSlice) (errs error) {
```
Comment (dmitryax) [inline pkg/pdatatest/ptracetest/traces.go:303]: Please don't expose it, we don't expose compare functions for slices here

```suggestion
func compareSpanEventSlice(expected, actual ptrace.SpanEventSlice) (errs error) {
```

File: pkg/pdatatest/ptracetest/traces.go [hunk 4]
```diff
@@ -301,3 +297,84 @@ func CompareSpan(expected, actual ptrace.Span) error {
...
+		errs = multierr.Append(errs, fmt.Errorf("timestamp doesn't match expected: %d, actual: %d",
+			expected.Timestamp(), actual.Timestamp()))
+	}
+
+	return errs
+}
+
+// CompareLinks compares each part of two given SpanLinkSlice and returns
+// an error if they don't match. The error describes what didn't match.
+func CompareLinks(expected, actual ptrace.SpanLinkSlice) (errs error) {
```
Comment (dmitryax) [inline pkg/pdatatest/ptracetest/traces.go:341]: Same, please don't expose it

File: pkg/pdatatest/ptracetest/traces.go [hunk 5]
```diff
@@ -301,3 +297,84 @@ func CompareSpan(expected, actual ptrace.Span) error {
...
+		ee := expected.At(i)
+		ae := actual.At(i)
+		errs = multierr.Append(errs, internal.AddErrPrefix(fmt.Sprintf(`event "%s"`, ee.Name()), CompareEvent(ae, ee)))
+	}
+	return errs
+}
+
+// CompareEvent compares each part of two given SpanEvent and returns
+// an error if they don't match. The error describes what didn't match.
+func CompareEvent(expected, actual ptrace.SpanEvent) error {
```
Comment (dmitryax) [inline pkg/pdatatest/ptracetest/traces.go:320]: Let's call the functions consistently based on the argument type
```suggestion
func CompareSpanEvent(expected, actual ptrace.SpanEvent) error {
```

File: pkg/pdatatest/ptracetest/traces.go [hunk 6]
```diff
@@ -301,3 +297,84 @@ func CompareSpan(expected, actual ptrace.Span) error {
...
+		el := expected.At(i)
+		al := actual.At(i)
+		errs = multierr.Append(errs, internal.AddErrPrefix(fmt.Sprintf(`link "%s"`, el.SpanID()), CompareLink(al, el)))
+	}
+	return errs
+}
+
+// CompareLink compares each part of two given SpanLink and returns
+// an error if they don't match. The error describes what didn't match.
+func CompareLink(expected, actual ptrace.SpanLink) error {
```
Comment (dmitryax) [inline pkg/pdatatest/ptracetest/traces.go:358]:
```suggestion
func CompareSpanLink(expected, actual ptrace.SpanLink) error {
```

**Review by dmitryax — APPROVED** [review]
LGTM. Thanks, @fatsheep9146!

**Comment by fatsheep9146** [comment]
> LGTM. Thanks, @fatsheep9146!

@dmitryax Can I volunteer myself as owner of pkg/pdatatest? I'd love to keep reviewing and contributing to this.

**Comment by dmitryax** [comment]
@fatsheep9146  Sure. Feel free to submit a PR

**Pull request merged**